### DISCUSSION TITLE
Auth: session() vs user()

### DISCUSSION THREAD

**Discussion by taylorhlabs** [supabase/supabase#4400] [Questions] [answered]
I've looked over the docs but haven't found much info on when to use `supabase.auth.session()` vs `supabase.auth.user()`. I've seen both used in similar examples for checking user data. For instance, when just checking if the user is logged in and then either showing some user details or a login screen, is one method preferred over the other?

**Comment by michaelpumo** [comment]
I'd also like to know this! Did you find an answer? Thanks

**Comment by Hum4n01d** [comment]
It seems that the two are the same, is there any official word on this?

**Comment by ComputelessComputer** [comment]
Note: This is a guess.

According to the official doc, the `user()` method returns the user data, **if there is a logged in user.**
Which I assume to be handled by a cookie since the doc states using it for server-side management.

https://supabase.com/docs/reference/javascript/auth-user
> Inside a browser context, user() will return the user data, if there is a logged in user.
For server-side management, you can get a user through auth.api.getUserByCookie()

So probably `user()` uses `session()` in order to retrieve the session data as a cookie, and then use the cookie to get the actual user data, if logged in.

**Comment by J0** [comment, marked as answer]
In Supabase-js  V1:

`session()` is not guaranteed to contain a user so if you wanted to get the logged in user you would use `user()`  to check if there's a logged in user. 

For Supabase-js v2:

User is contained on the session.

`getUser(jwt?:string)` takes in a jwt and is mainly intended for use Server Side where one might not have access to the Session data via cookie. In this scenario, we can make use of the JWT to get the user.

Let us know if there are further queries!

**Reply by liaujianjie** [reply]
Suppose I would like to check if a user is authenticated on the server side, is there any guidance on which I should use for Supabase JS v2?

Specifically, which is more accurate? If they are both equally accurate, then which is more efficient (e.g. less external API calls)?

**Reply by liaujianjie** [reply]
Oops, I've just  found it in [the docs](https://supabase.com/docs/reference/javascript/auth-getuser):

> `getUser()` should be used only when you require the most current user data. For faster results, `getSession().session.user` is recommended.

**Reply by enzzzooo** [reply]
> For Supabase-js v2:
> 
> User is contained on the session.
> 
> `getUser(jwt?:string)` takes in a jwt and is mainly intended for use Server Side where one might not have access to the Session data via cookie. In this scenario, we can make use of the JWT to get the user.
> 
> Let us know if there are further queries!

How would you add the user’s jwt to the function call `getUser(jwt?:string)`?

**Comment by jasoncarrillo0** [comment]
The docs are somewhat confusing on this matter. 

For example, the getSession docs say:
> Since the unencoded session data is retrieved from the local storage medium, do not rely on it as a source of trusted data on the server. It could be tampered with by the sender. If you need verified, trustworthy user data, call [getUser](https://supabase.com/docs/reference/javascript/auth-getuser) instead.

and then the getUser docs say:
>Should be used only when you require the most current user data. For faster results, getSession().session.user is recommended.

So it's saying that `getSession()` is recommended for non-current user data, while at the same time saying that `getSession()` is completely unreliable for user data and could be tampered with. 

It makes me wonder, what is the point of even having a `getSession()` function? 

For use only on the frontend I suppose. And the backend should only ever use `getSession()`?

**Reply by charislam** [reply]
Yes, you can use `getSession` on the frontend to speed things up because the frontend is never secure from user tampering anyway.
And on the backend you should only use `getUser` (I assume that's what you intended to write in your post 😄)

I believe the latest version of the Supabase helper packages/client SDK now include a warning about this (not 100% sure if it's released yet, but it's been added in main).

(Good point about the docs being confusing, we'll update them.)

**Reply by jasoncarrillo0** [reply]
Thanks, @charislam. It should be noted that the base nextjs supabase starter with ssr uses `getSession()` to refresh a session in global nextjs middleware. Also the [auth helpers docs say it too.](https://supabase.com/docs/guides/auth/auth-helpers/nextjs#managing-session-with-middleware). It appears the only place that has the right info about not trusting `getSession()` in middleware is a doc page for [setting up server-side auth with nextjs](https://supabase.com/docs/guides/auth/server-side/nextjs).

**Reply by charislam** [reply]
Thanks for finding all the places we've missed! I've opened a [PR](https://github.com/supabase/supabase/pull/21436) to fix it.

**Comment by viantirreau** [comment]
Aligned with @jasoncarrillo0's question, when using the `supabase` client specifically in Next.js, `getUser()` is already called in the suggested middleware code. Is it advisable, from a security standpoint, to also call `getUser()` inside private user routes, provided that the middleware should have been called beforehand, or is it enough to stick to `getSession()` for lower latency in those cases?

**Reply by viantirreau** [reply]
An update on this.
After deploying the change suggested by the [docs](https://github.com/supabase/supabase/blob/5a2600fca918750d545359c077bff653a893d350/apps/docs/content/guides/auth/server-side/nextjs.mdx#L258), i.e. replacing
```js
await supabase.auth.getSession()
```
by
```js
await supabase.auth.getUser()
```
in the `middleware.ts` file, I noticed a huge increase in the amount of auth requests hitting the supabase instance, and the culprit seems to be link prefetching. It makes a lot of sense: beforehand, we only needed to read the cookies to assert that the session was "valid", so it could be done completely server-side, without calls to supabase as an oracle. Now, with `getUser()`, we need to validate that the cookies correspond to an actually valid session.

As suggested in [this issue's answer](https://github.com/vercel/next.js/discussions/37736#discussioncomment-7274460), a workaround could be to disable the middleware when the request is a prefetch call:

```js
export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     * Also, disable the middleware on prefetch calls.
     * Feel free to modify this pattern to include more paths.
     */
    {
      source: '/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)',
      missing: [
        { type: 'header', key: 'next-router-prefetch' },
        { type: 'header', key: 'purpose', value: 'prefetch' },
      ]
    }
  ]
}
```
